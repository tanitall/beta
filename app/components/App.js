import React, { Component } from "react";
import { connect } from "react-redux";
import SplitPane from "react-split-pane";
import Modal from "react-modal";
import spinner from "../img/spinner.png";

const styles = {
  overlay: {
    position: "fixed",
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    backgroundColor: "rgba(0, 0, 0, 0.75)"
  },
  content: {
    margin: "100px auto 0",
    padding: "30px 30px 30px 30px",
    border: "thick solid #222",
    background: "rgba(12, 12, 14, 1)",
    borderRadius: "20px",
    top: "100px",
    height: 261,
    width: 559,
    left: "100px",
    right: "100px",
    bottom: "100px",
    boxShadow: "0px 10px 44px rgba(0, 0, 0, 0.45)"
  }
};

const StatusMessage = ({ status, statusMessage }) => {
  let message = null;
  if (status === true) {
    message = (
      <Modal
        isOpen={true}
        closeTimeoutMS={5}
        style={styles}
        contentLabel="Modal"
      >
        <div>
          <div className="center modal-alert">
            <img src={spinner} alt="" className="spinner" />
          </div>
          <div className="center modal-alert">
            <strong>{statusMessage}</strong>
          </div>
        </div>
      </Modal>
    );
  } else if (status === false) {
    message = (
      <Modal
        isOpen={true}
        closeTimeoutMS={5}
        style={styles}
        contentLabel="Modal"
      >
        <div>
          <div className="center modal-alert">
            <img src={spinner} alt="" className="spinner" />
          </div>
          <div className="center modal-alert">
            <strong>{statusMessage}</strong>
          </div>
        </div>
      </Modal>
    );
  }
  return message;
};

class App extends Component {
  constructor() {
    super();
    this.state = {
      modalIsOpen: false
    };

    this.closeModal = this.closeModal.bind(this);
  }
  closeModal() {
    this.setState({ modalIsOpen: false });
  }
  render() {
    return (
      <div id="pageWrapper">
        <StatusMessage
          status={this.props.status}
          statusMessage={this.props.statusMessage}
        />
        <div>{this.props.children}</div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  status: state.transactions.success,
  statusMessage: state.transactions.message
});

App = connect(mapStateToProps)(App);

export default App;
