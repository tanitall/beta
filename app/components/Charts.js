import React, { Component } from "react";
import { connect } from "react-redux";
import ReactTooltip from "react-tooltip";
import { Line } from "react-chartjs-2";
import axios from "axios";
import _ from "lodash";
import moment from "moment";
import neoLogo from "../images/neo.png";
import { Link } from "react-router";

const api = val => {
  return `https://min-api.cryptocompare.com/data/histohour?fsym=${
    val
  }&tsym=USD&limit=72&aggregate=3&e=CCCAGG`;
};

class Charts extends Component {
  constructor(props) {
    super(props);
    this.state = {
      neoData: [],
      gasData: [],
      btcData: [],
      dashData: [],
      ltcData: [],
      ethData: [],
      rpxData: [],
      open: "--",
      high: "--",
      low: "--"
    };
  }

  async componentDidMount() {
    await this.getGasData();
    await this.getNeoData();
    await this.getBtcData();
    await this.getLtcData();
    await this.getDashData();
    await this.getEthData();
    await this.getRpxData();
  }

  async getGasData() {
    try {
      let req = await axios.get(api("GAS"));
      let data = req.data.Data;
      this.setState({ gasData: data });
    } catch (error) {
      console.log(error);
    }
  }

  async getNeoData() {
    try {
      let req = await axios.get(api("NEO"));
      let data = req.data.Data;
      this.setState({ neoData: data });
      this.setState({ ...data[95] });
    } catch (error) {
      console.log(error);
    }
  }

  async getBtcData() {
    try {
      let req = await axios.get(api("BTC"));
      let data = req.data.Data;
      this.setState({ btcData: data });
    } catch (error) {
      console.log(error);
    }
  }

  async getLtcData() {
    try {
      let req = await axios.get(api("LTC"));
      let data = req.data.Data;
      this.setState({ ltcData: data });
    } catch (error) {
      console.log(error);
    }
  }

  async getEthData() {
    try {
      let req = await axios.get(api("ETH"));
      let data = req.data.Data;
      this.setState({ ethData: data });
    } catch (error) {
      console.log(error);
    }
  }

  async getDashData() {
    try {
      let req = await axios.get(api("DASH"));
      let data = req.data.Data;
      this.setState({ dashData: data });
    } catch (error) {
      console.log(error);
    }
  }

  async getRpxData() {
    try {
      let req = await axios.get(api("RPX"));
      let data = req.data.Data;
      this.setState({ rpxData: data });
    } catch (error) {
      console.log(error);
    }
  }

  render() {
    const neoPrices = _.map(this.state.neoData, "close");
    const neoHours = _.map(this.state.neoData, "time");

    const convertedTime = neoHours.map(val => moment.unix(val).format("LLL"));

    const ltcPrices = _.map(this.state.ltcData, "close");
    const btcPrices = _.map(this.state.btcData, "close");
    const dashPrices = _.map(this.state.dashData, "close");
    const ethPrices = _.map(this.state.ethData, "close");
    const rpxPrices = _.map(this.state.rpxData, "close");
    const gasPrices = _.map(this.state.gasData, "close");

    const data = canvas => {
      let ctx = canvas.getContext("2d");
      let gradientStroke = ctx.createLinearGradient(500, 0, 100, 0);
      gradientStroke.addColorStop(0, "#7ED321");
      gradientStroke.addColorStop(1, "#7ED321");

      let gradientFill = ctx.createLinearGradient(0, 0, 0, 240);
      gradientFill.addColorStop(0, "rgba(68,147,33,0.8)");
      gradientFill.addColorStop(1, "rgba(68,147,33,0)");
      const gradient = ctx.createLinearGradient(0, 0, 100, 0);

      let gasGradientStroke = ctx.createLinearGradient(500, 0, 100, 0);
      gasGradientStroke.addColorStop(0, "#9013FE");
      gasGradientStroke.addColorStop(1, "#9013FE");

      let gasGradientFill = ctx.createLinearGradient(0, 0, 0, 240);
      gasGradientFill.addColorStop(0, "rgba(144,147,254, 1)");
      gasGradientFill.addColorStop(1, "rgba(144,147,254, 0)");

      let ltcGradientStroke = ctx.createLinearGradient(500, 0, 100, 0);
      ltcGradientStroke.addColorStop(0, "#ececec");
      ltcGradientStroke.addColorStop(1, "#ececec");

      let ltcGradientFill = ctx.createLinearGradient(0, 0, 0, 240);
      ltcGradientFill.addColorStop(0, "rgba(255,255,255, 0.5)");
      ltcGradientFill.addColorStop(1, "rgba(255,255,255, 0)");

      let btcGradientStroke = ctx.createLinearGradient(500, 0, 100, 0);
      btcGradientStroke.addColorStop(0, "#ffc000");
      btcGradientStroke.addColorStop(1, "#ffc000");

      let btcGradientFill = ctx.createLinearGradient(0, 0, 0, 240);
      btcGradientFill.addColorStop(0, "rgba(229,172,0, 0.5)");
      btcGradientFill.addColorStop(1, "rgba(229,172,0, 0)");

      let ethGradientStroke = ctx.createLinearGradient(500, 0, 100, 0);
      ethGradientStroke.addColorStop(0, "#646464");
      ethGradientStroke.addColorStop(1, "#646464");

      let ethGradientFill = ctx.createLinearGradient(0, 0, 0, 240);
      ethGradientFill.addColorStop(0, "rgba(175,175,175, 0.5)");
      ethGradientFill.addColorStop(1, "rgba(175,175,175, 0)");

      let rpxGradientStroke = ctx.createLinearGradient(500, 0, 100, 0);
      rpxGradientStroke.addColorStop(0, "#C60307");
      rpxGradientStroke.addColorStop(1, "#C60307");

      let rpxGradientFill = ctx.createLinearGradient(0, 0, 0, 240);
      rpxGradientFill.addColorStop(0, "rgba(169,3,41, 0.5)");
      rpxGradientFill.addColorStop(1, "rgba(169,3,41, 0)");

      let dashGradientStroke = ctx.createLinearGradient(500, 0, 100, 0);
      dashGradientStroke.addColorStop(0, "#005aff");
      dashGradientStroke.addColorStop(1, "#005aff");

      let dashGradientFill = ctx.createLinearGradient(0, 0, 0, 240);
      dashGradientFill.addColorStop(0, "rgba(0,90,255, 0.5)");
      dashGradientFill.addColorStop(1, "rgba(0,90,255, 0)");

      return {
        labels: convertedTime,
        datasets: [
          {
            label: "GAS",
            fill: true,
            lineTension: 0.25,
            backgroundColor: gasGradientFill,
            borderColor: gasGradientStroke,
            borderCapStyle: "butt",
            borderDash: [],
            borderDashOffset: 0.0,
            borderJoinStyle: "miter",
            pointBorderWidth: 1,
            pointHoverRadius: 3,
            pointHoverBorderWidth: 0,
            pointBorderColor: gasGradientStroke,
            pointBackgroundColor: gasGradientStroke,
            pointHoverBackgroundColor: gasGradientStroke,
            pointHoverBorderColor: gasGradientStroke,
            pointHitRadius: 3,
            pointRadius: 0,
            data: gasPrices
          },
          {
            label: "NEO",
            fill: true,
            lineTension: 0.25,
            backgroundColor: gradientFill,
            borderColor: gradientStroke,
            borderCapStyle: "butt",
            borderDash: [],
            borderDashOffset: 0.0,
            borderJoinStyle: "miter",
            pointBorderWidth: 1,
            pointHoverRadius: 3,
            pointHoverBorderWidth: 0,
            pointBorderColor: gradientStroke,
            pointBackgroundColor: gradientStroke,
            pointHoverBackgroundColor: gradientStroke,
            pointHoverBorderColor: gradientStroke,
            pointHitRadius: 3,
            pointRadius: 0,
            data: neoPrices
          },
          {
            label: "BTC",
            fill: true,
            hidden: true,
            lineTension: 0.25,
            backgroundColor: btcGradientFill,
            borderColor: btcGradientStroke,
            borderCapStyle: "butt",
            borderDash: [],
            borderDashOffset: 0.0,
            borderJoinStyle: "miter",
            pointBorderWidth: 1,
            pointHoverRadius: 3,
            pointHoverBorderWidth: 0,
            pointBorderColor: btcGradientStroke,
            pointBackgroundColor: btcGradientStroke,
            pointHoverBackgroundColor: btcGradientStroke,
            pointHoverBorderColor: btcGradientStroke,
            pointHitRadius: 3,
            pointRadius: 0,
            data: btcPrices
          },
          {
            label: "DASH",
            fill: true,
            hidden: true,
            lineTension: 0.25,
            backgroundColor: dashGradientFill,
            borderColor: dashGradientStroke,
            borderCapStyle: "butt",
            borderDash: [],
            borderDashOffset: 0.0,
            borderJoinStyle: "miter",
            pointBorderWidth: 1,
            pointHoverRadius: 3,
            pointHoverBorderWidth: 0,
            pointBorderColor: dashGradientStroke,
            pointBackgroundColor: dashGradientStroke,
            pointHoverBackgroundColor: dashGradientStroke,
            pointHoverBorderColor: dashGradientStroke,
            pointHitRadius: 3,
            pointRadius: 0,
            data: dashPrices
          },
          {
            label: "ETH",
            fill: true,
            hidden: true,
            lineTension: 0.25,
            backgroundColor: ethGradientFill,
            borderColor: ethGradientStroke,
            borderCapStyle: "butt",
            borderDash: [],
            borderDashOffset: 0.0,
            borderJoinStyle: "miter",
            pointBorderWidth: 1,
            pointHoverRadius: 3,
            pointHoverBorderWidth: 0,
            pointBorderColor: ethGradientStroke,
            pointBackgroundColor: ethGradientStroke,
            pointHoverBackgroundColor: ethGradientStroke,
            pointHoverBorderColor: ethGradientStroke,
            pointHitRadius: 3,
            pointRadius: 0,
            data: ethPrices
          },
          {
            label: "LTC",
            fill: true,
            hidden: true,
            lineTension: 0.25,
            backgroundColor: ltcGradientFill,
            borderColor: ltcGradientStroke,
            borderCapStyle: "butt",
            borderDash: [],
            borderDashOffset: 0.0,
            borderJoinStyle: "miter",
            pointBorderWidth: 1,
            pointHoverRadius: 3,
            pointHoverBorderWidth: 0,
            pointBorderColor: ltcGradientStroke,
            pointBackgroundColor: ltcGradientStroke,
            pointHoverBackgroundColor: ltcGradientStroke,
            pointHoverBorderColor: ltcGradientStroke,
            pointHitRadius: 3,
            pointRadius: 0,
            data: ltcPrices
          },
          {
            label: "RPX",
            fill: true,
            hidden: true,
            lineTension: 0.25,
            backgroundColor: rpxGradientFill,
            borderColor: rpxGradientStroke,
            borderCapStyle: "butt",
            borderDash: [],
            borderDashOffset: 0.0,
            borderJoinStyle: "miter",
            pointBorderWidth: 1,
            pointHoverRadius: 3,
            pointHoverBorderWidth: 0,
            pointBorderColor: rpxGradientStroke,
            pointBackgroundColor: rpxGradientStroke,
            pointHoverBackgroundColor: rpxGradientStroke,
            pointHoverBorderColor: rpxGradientStroke,
            pointHitRadius: 3,
            pointRadius: 0,
            data: rpxPrices
          }
        ]
      };
    };
    return (
      <div>

        <div className="settings-panel">
          <div className="row">
            <div className="col-xs-12">
              <div className="col-xs-4">
                <img
                  src={neoLogo}
                  alt=""
                  width="34"
                  className="neo-logo logobounce"
                />
                <h3 className="neo-dash-price">NEO Price</h3>
              </div>
              <div className="col-xs-6"/>
              <div className="col-xs-2">
                <select
                  name="select-profession"
                  id="dash-price-select"
                  className=""
                >
                  <option defaultValue selected="selected">
                    USD
                  </option>
                </select>
              </div>
              <Line
                data={data}
                width={600}
                height={300}
                options={{
                  maintainAspectRatio: true,
                  layout: { padding: { left: 0, right: 0, top: 0, bottom: 0 } },
                  scales: {
                    xAxes: [
                      {
                        type: "time",
                        position: "bottom",
                        id: "x-axis-0",
                        categoryPercentage: 0.25,
                        barPercentage: 0.25,
                        gridLines: { color: "rgba(255, 255, 255, 0.04)" }
                      }
                    ],
                    yAxes: [
                      {
                        gridLines: { color: "rgba(255, 255, 255, 0.04)" }
                      }
                    ]
                  },
                  legend: { position: "bottom" }
                }}
              />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  neo: state.wallet.Neo,
  gas: state.wallet.Gas,
  address: state.account.address,
  net: state.metadata.network,
  price: state.wallet.price
});

Charts = connect(mapStateToProps)(Charts);

export default Charts;
