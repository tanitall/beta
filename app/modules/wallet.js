// Constants
export const SET_BALANCE = "SET_BALANCE";
export const SET_MARKET_PRICE = "SET_MARKET_PRICE";
export const RESET_PRICE = "RESET_PRICE";
export const SET_TRANSACTION_HISTORY = "SET_TRANSACTION_HISTORY";

// Actions
export function setBalance(neo, gas, price, combined, gasPrice) {
  return {
    type: SET_BALANCE,
    Neo: neo,
    Gas: gas,
    price: price,
    combined: combined,
    gasPrice: gasPrice
  };
}

export function setMarketPrice(price) {
  return {
    type: SET_MARKET_PRICE,
    price: price
  };
}

export function resetPrice() {
  return {
    type: RESET_PRICE
  };
}

export function setTransactionHistory(transactions) {
  return {
    type: SET_TRANSACTION_HISTORY,
    transactions
  };
}

// reducer for wallet account balance
export default (
  state = {
    Neo: 0,
    Gas: 0,
    transactions: [],
    price: "--",
    combined: "--",
    gasPrice: "--"
  },
  action
) => {
  switch (action.type) {
    case SET_BALANCE:
      return {
        ...state,
        Neo: action.Neo,
        Gas: action.Gas,
        price: action.price,
        combined: action.combined,
        gasPrice: action.gasPrice
      };
    case RESET_PRICE:
      return { ...state, price: "--" };
    case SET_MARKET_PRICE: //current market price action type
      let currentPrice;
      if (action.price !== undefined) {
        currentPrice = action.price;
      } else {
        currentPrice = "--";
      }
      return { ...state, price: currentPrice };
    case SET_TRANSACTION_HISTORY:
      return { ...state, transactions: action.transactions };
    default:
      return state;
  }
};
