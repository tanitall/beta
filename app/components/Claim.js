import React, { Component } from "react";
import { connect } from "react-redux";
import { setClaimRequest, disableClaim } from "../modules/claim";
import { sendEvent, clearTransactionEvent } from "../modules/transactions";
import { doClaimAllGas, doSendAsset } from "neon-js";
import ReactTooltip from "react-tooltip";
import { log } from "../util/Logs";

// wrap claiming with notifications

const doClaimNotify = (dispatch, net, selfAddress, wif) => {
  log(net, "CLAIM", selfAddress, { info: "claim all gas" });
  doClaimAllGas(net, wif).then(response => {
    if (response.result === true) {
      dispatch(
        sendEvent(
          true,
          "Your GAS claim was successful! Your balance will update once the blockchain has processed it."
        )
      );
      setTimeout(() => dispatch(disableClaim(false)), 3000);
    } else {
      dispatch(sendEvent(false, "Sorry. Claim failed. Please try again."));
    }
    setTimeout(() => dispatch(clearTransactionEvent()), 5000);
  });
};

// To initiate claim, first send all Neo to own address, the set claimRequest state
// When new claims are available, this will trigger the claim
const doGasClaim = (dispatch, net, wif, selfAddress, ans) => {
  // if no neo in account, no need to send to self first
  if (ans === 0) {
    doClaimNotify(dispatch, net, selfAddress, wif);
  } else {
    dispatch(sendEvent(true, "Sending 0 Neo in order to claim your GAS..."));
    log(net, "SEND", selfAddress, {
      to: selfAddress,
      amount: ans,
      asset: "NEO"
    });
    doSendAsset(net, selfAddress, wif, "Neo", ans).then(response => {
      if (response.result === undefined || response.result === false) {
        dispatch(
          sendEvent(false, "Oops! Transaction failed. Please try again.")
        );
      } else {
        dispatch(sendEvent(true, "Waiting for the transaction to clear..."));
        dispatch(setClaimRequest(true));
        dispatch(disableClaim(true));
      }
    });
  }
};

class Claim extends Component {
  componentDidUpdate = () => {
    if (
      this.props.claimRequest === true &&
      this.props.claimWasUpdated == true
    ) {
      this.props.dispatch(setClaimRequest(false));
      doClaimNotify(
        this.props.dispatch,
        this.props.net,
        this.props.address,
        this.props.wif
      );
    }
  };

  render = () => {
    let renderButton;
    const doClaim = () =>
      doGasClaim(
        this.props.dispatch,
        this.props.net,
        this.props.wif,
        this.props.address,
        this.props.neo
      );
    if (this.props.disableClaimButton === false) {
      renderButton = (
        <div id="gas-button" onClick={doClaim}>
          <span className="gas-claim">
            Claim Gas<br />
            {this.props.claimAmount}
          </span>
        </div>
      );
    } else {
      renderButton = (
        <div>
          <div id="gas-button" data-tip data-for="claimTip" className="">
            <span className="gas-claim">
              Claim Gas<br />
              {this.props.claimAmount}
            </span>
          </div>
          <ReactTooltip
            className="solidTip"
            id="claimTip"
            place="bottom"
            type="dark"
            effect="solid"
          >
            <span>You can claim GAS once every 5 minutes</span>
          </ReactTooltip>
        </div>
      );
    }
    return <div id="claim">{renderButton}</div>;
  };
}

const mapStateToProps = state => ({
  claimAmount: state.claim.claimAmount,
  claimRequest: state.claim.claimRequest,
  claimWasUpdated: state.claim.claimWasUpdated,
  disableClaimButton: state.claim.disableClaimButton,
  wif: state.account.wif,
  address: state.account.address,
  net: state.metadata.network,
  neo: state.wallet.Neo
});

Claim = connect(mapStateToProps)(Claim);

export default Claim;
