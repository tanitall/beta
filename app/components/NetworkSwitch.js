import React, { Component } from "react";
import { connect } from "react-redux";
import {
  getBalance,
  getTransactionHistory,
  getClaimAmounts,
  getWalletDBHeight,
  getAPIEndpoint
} from "neon-js";
import { setClaim } from "../modules/claim";
import { setBlockHeight, setNetwork } from "../modules/metadata";
import {
  setBalance,
  setMarketPrice,
  resetPrice,
  setTransactionHistory
} from "../modules/wallet";
import { version } from "../../package.json";
import { sendEvent, clearTransactionEvent } from "../modules/transactions";
import axios from "axios";

let intervals = {};

let netSelect;

// https://bittrex.com/api/v1.1/public/getmarkethistory?market=BTC-NEO

// putting this back in wallet, does not belong in neon-js
export const getMarketPriceUSD = amount => {
  return axios
    .get("https://bittrex.com/api/v1.1/public/getticker?market=USDT-NEO")
    .then(response => {
      let lastUSDNEO = Number(response.data.result.Last);
      return "$" + (lastUSDNEO * amount).toFixed(2).toString();
    });
};

const getGasPrice = async gasVal => {
  try {
    let gas = await axios.get("https://api.coinmarketcap.com/v1/ticker/gas/");
    gas = gas.data[0].price_usd;
    const value = gasVal * gas;
    return value;
  } catch (error) {
    console.log(error);
  }
};

// TODO: this is being imported by Balance.js, maybe refactor to helper file

const initiateGetBalance = (dispatch, net, address) => {
  syncTransactionHistory(dispatch, net, address);
  syncAvailableClaim(dispatch, net, address);
  syncBlockHeight(dispatch, net);
  // checkVersion(dispatch, net);
  return getBalance(net, address)
    .then(resultBalance => {
      return getMarketPriceUSD(resultBalance.Neo)
        .then(async resultPrice => {
          if (resultPrice === undefined || resultPrice === null) {
            dispatch(setBalance(resultBalance.Neo, resultBalance.Gas, "--"));
          } else {
            let gasPrice = await getGasPrice(resultBalance.Gas);
            dispatch(
              setBalance(
                resultBalance.Neo,
                resultBalance.Gas,
                resultPrice,
                "--",
                gasPrice
              )
            );
          }
          return true;
        })
        .catch(e => {
          dispatch(setBalance(resultBalance.Neo, resultBalance.Gas, "--"));
          console.log("something went wrong");
        });
    })
    .catch(result => {
      // If API dies, still display balance
    });
};

const syncAvailableClaim = (dispatch, net, address) => {
  getClaimAmounts(net, address).then(result => {
    //claimAmount / 100000000
    dispatch(setClaim(result.available, result.unavailable));
  });
};

const syncBlockHeight = (dispatch, net) => {
  getWalletDBHeight(net).then(blockHeight => {
    dispatch(setBlockHeight(blockHeight));
  });
};

const syncTransactionHistory = (dispatch, net, address) => {
  getTransactionHistory(net, address).then(transactions => {
    let txs = [];
    for (let i = 0; i < transactions.length; i++) {
      if (transactions[i].neo_sent === true) {
        txs = txs.concat([
          {
            type: "NEO",
            amount: transactions[i].NEO,
            txid: transactions[i].txid,
            block_index: transactions[i].block_index
          }
        ]);
      }
      if (transactions[i].gas_sent === true) {
        txs = txs.concat([
          {
            type: "GAS",
            amount: transactions[i].GAS,
            txid: transactions[i].txid,
            block_index: transactions[i].block_index
          }
        ]);
      }
    }
    dispatch(setTransactionHistory(txs));
  });
};

const resetBalanceSync = (dispatch, net, address) => {
  if (intervals.balance !== undefined) {
    clearInterval(intervals.balance);
  }
  intervals.balance = setInterval(() => {
    initiateGetBalance(dispatch, net, address);
  }, 30000);
};

const toggleNet = (dispatch, net, address) => {
  let newNet;
  if (net === "MainNet") {
    newNet = "TestNet";
  } else {
    newNet = "MainNet";
  }
  dispatch(setNetwork(newNet));
  resetBalanceSync(dispatch, newNet, address);
  if (address !== null) {
    initiateGetBalance(dispatch, newNet, address);
  }
};

class NetworkSwitch extends Component {
  componentDidMount = () => {
    resetBalanceSync(this.props.dispatch, this.props.net, this.props.address);
  };
  render = () => (
    <div id="network">
      <div className="dash-icon-bar">
        <div className="icon-border">
          <span className="glyphicon glyphicon-exclamation-sign" />
        </div>
        <span className="transparent">Running on </span>
        <span className="netName">{this.props.net}</span>
      </div>
    </div>
  );
}

const mapStateToProps = state => ({
  net: state.metadata.network,
  address: state.account.address
});

NetworkSwitch = connect(mapStateToProps)(NetworkSwitch);

export {
  NetworkSwitch,
  initiateGetBalance,
  syncTransactionHistory,
  intervals,
  resetBalanceSync
};
