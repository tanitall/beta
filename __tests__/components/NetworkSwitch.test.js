import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import networkReducer, { setNetwork, SET_HEIGHT, SET_NETWORK } from '../../app/modules/metadata';
import { SET_CLAIM } from '../../app/modules/claim';
import { SET_BALANCE, SET_TRANSACTION_HISTORY } from '../../app/modules/wallet';
import { NetworkSwitch, resetBalanceSync } from '../../app/components/NetworkSwitch';

// TODO research how to move the axios mock code which is repeated in NetworkSwitch to a helper or config file
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { version } from '../../package.json';

const axiosMock = new MockAdapter(axios);
axiosMock
  .onGet('http://testnet-api.wallet.cityofzion.io/v2/version')
  .reply(200, { version });
axiosMock
  .onGet('https://bittrex.com/api/v1.1/public/getticker?market=USDT-NEO')
  .reply(200, { result: { Last: 24.50 } });

jest.mock('neon-js');

const initialState = {
  account: {
    address: 'AWy7RNBVr9vDadRMK9p7i7Z1tL7GrLAxoh'
  },
  metadata: {
    network: 'MainNet'
  }
};

const setup = (state = initialState) => {
  const store = configureStore()(state);
  const wrapper = shallow(<NetworkSwitch store={store} />);

  return {
    store,
    wrapper
  };
};

describe('NetworkSwitch', () => {
  test('renders without crashing', (done) => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
    done();
  });

  test('calls resetBalanceSync after rendering', (done) => {
    const { wrapper, store } = setup();
    const state = store.getState();
    expect(store.getActions()[0]).toEqual(resetBalanceSync(store.dispatch, state.metadata.network, state.account.address));
    done();
  });

  test('correctly renders MainNet initially', (done) => {
    const { wrapper, store } = setup();
    const state = store.getState();
    expect(wrapper.dive().find('.netName').text()).toEqual(state.metadata.network);
    done();
  });

  test('switches to TestNet when clicked', (done) => {
    const { wrapper, store } = setup();
    const state = store.getState();
    const testNet = 'TestNet';
    const deepWrapper = wrapper.dive();
    expect(deepWrapper.find('.netName').text()).toEqual(state.metadata.network);
    const actionTypes = [
      SET_NETWORK,
      SET_HEIGHT,
      SET_CLAIM,
      SET_TRANSACTION_HISTORY,
      SET_BALANCE
    ]
    deepWrapper.find('.netName').simulate('click');
    setTimeout(() => {
      const actions = store.getActions();
      actions.forEach(action => {
        expect(actionTypes.indexOf(action.type) > -1).toEqual(true)
      });
      expect(actions.length).toEqual(5);
      done();
    }, 0)
  });
});
