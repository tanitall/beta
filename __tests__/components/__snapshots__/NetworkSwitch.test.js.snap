// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkSwitch renders without crashing 1`] = `
<NetworkSwitch
  address="AWy7RNBVr9vDadRMK9p7i7Z1tL7GrLAxoh"
  dispatch={[Function]}
  net="MainNet"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
